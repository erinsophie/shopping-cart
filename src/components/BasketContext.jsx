import { createContext, useContext, useState, useEffect, useRef } from 'react';

const BasketContext = createContext();

function useBasket() {
  return useContext(BasketContext);
}

function BasketProvider({ children, initialBasket = [] }) {
  const [showBasket, setShowBasket] = useState(false);
  const [basket, setBasket] = useState(initialBasket);
  const basketRef = useRef(basket);

  // when basket changes, set the current ref to be
  // the most recent state of the basket
  // this avoids discrepencies in the quantity
  // of products when addToBasket is clicked in succession

  console.log(basket)

  useEffect(() => {
    basketRef.current = basket;
  }, [basket]);

  function addToBasket(newProduct) {
    const currentBasket = basketRef.current;

    const existingItemIndex = currentBasket.findIndex(
      (item) => item.product.id === newProduct.id,
    );

    if (existingItemIndex !== -1) {
      const updatedBasket = [...currentBasket];
      updatedBasket[existingItemIndex].quantity += 1;
      setBasket(updatedBasket);
    } else {
      setBasket([...currentBasket, { product: newProduct, quantity: 1 }]);
    }
  }

  function updateQuantity(basketItem, event) {
    const currentBasket = basketRef.current;
    const indexOfItem = basket.indexOf(basketItem);
    const updatedBasket = [...currentBasket];

    if (event.target.textContent === '-') {
      if (updatedBasket[indexOfItem].quantity === 1) {
        deleteProduct(basketItem.product.id);
      } else {
        updatedBasket[indexOfItem].quantity -= 1;
        setBasket(updatedBasket);
      }
    } else if (event.target.textContent === '+') {
      updatedBasket[indexOfItem].quantity += 1;
      setBasket(updatedBasket);
    }
  }

  function getSubtotal(basket) {
    const subTotal = basket.reduce(
      (acc, item) => acc + item.product.price * item.quantity,
      0,
    );

    return subTotal.toFixed(2);
  }

  function getBasketAmount() {
    return basket.reduce((acc, item) => acc + item.quantity, 0);
  }

  function deleteProduct(productId) {
    const newBasket = basket.filter((item) => item.product.id !== productId);
    setBasket(newBasket);
  }

  function clearBasket() {
    setShowBasket(false);
    setBasket([]);
  }

  const providerValues = {
    basket,
    setBasket,
    showBasket,
    setShowBasket,
    addToBasket,
    getSubtotal,
    getBasketAmount,
    deleteProduct,
    updateQuantity,
    clearBasket,
  };

  return (
    <BasketContext.Provider value={providerValues}>
      {children}
    </BasketContext.Provider>
  );
}

export { useBasket, BasketProvider };
