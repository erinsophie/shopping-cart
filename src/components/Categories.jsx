import { useState, useEffect } from 'react';
import { capitaliseLetter } from '../utils/utils';
import { Link } from 'react-router-dom';

function Categories() {
  const [categories, setCategories] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  // fetch categories
  useEffect(() => {
    async function fetchCategories() {
      try {
        const response = await fetch(
          'https://fakestoreapi.com/products/categories',
        );
        if (!response.ok) {
          throw new Error('Network response was not ok ' + response.statusText);
        }
        const data = await response.json();
        setCategories(data);
        setError(null);
      } catch (error) {
        setError(error.message);
        setCategories(null);
      } finally {
        setLoading(false);
      }
    }

    fetchCategories();
  }, []);

  // display error
  if (error) return <p>Error: {error}</p>;

  return (
    <div className="flex justify-center gap-7 p-3 md:p-8">
      {loading ? (
        <p>Loading...</p>
      ) : (
        <div className="flex gap-5 md:gap-16">
          {categories.map((category, index) => (
            <Link to={`/${category}`} key={index} aria-label={category}>
              <p className="text-center transform transition-transform ease-in-out duration-200 hover:scale-110">
                {capitaliseLetter(category)}
              </p>
            </Link>
          ))}
        </div>
      )}
    </div>
  );
}

export default Categories;
