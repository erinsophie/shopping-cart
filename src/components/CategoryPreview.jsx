function CategoryPreview({ src, name }) {
  return (
    <div className="flex flex-col gap-2 items-center transform transition-transform ease-in-out duration-200 hover:scale-110 cursor-pointer">
      <div className="w-16 h-16 rounded-full overflow-hidden hover:shadow-xl md:w-24 md:h-24 lg:w-36 lg:h-36">
        <img src={src} alt={name} className="object-cover w-full h-full"></img>
      </div>
      <p className='text-xs text-center lg:text-base'>{name}</p>
    </div>
  );
}

export default CategoryPreview;
