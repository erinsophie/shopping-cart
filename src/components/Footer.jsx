function Footer() {
  return (
    <footer className="border-t border-black font-light text-xs md:text-sm lg:text-base">
      <div className="flex gap-6 p-6 bg-footerBlue md:justify-evenly md:p-10">
        <div>
          <p className="font-medium mb-5">Shop</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Gift cards</li>
            <li className="hover:underline">Sitemap</li>
            <li className="hover:underline">Blog</li>
          </ul>
        </div>
        <div>
          <p className="font-medium mb-5">Sell</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Teams</li>
            <li className="hover:underline">Forums</li>
          </ul>
        </div>
        <div>
          <p className="font-medium mb-5">About</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Policies</li>
            <li className="hover:underline">Investors</li>
            <li className="hover:underline">Careers</li>
            <li className="hover:underline">Press</li>
          </ul>
        </div>
        <div>
          <p className="font-medium mb-5">Help</p>
          <ul className="cursor-pointer flex flex-col gap-3">
            <li className="hover:underline">Help centre</li>
            <li className="hover:underline">Privacy Settings</li>
          </ul>

          <div className="flex gap-5 text-sm mt-10 md:text-2xl lg:text-3xl">
            <i className="fa-brands fa-instagram"></i>
            <i className="fa-brands fa-facebook"></i>
            <i className="fa-brands fa-pinterest"></i>
            <i className="fa-brands fa-twitter"></i>
            <i className="fa-brands fa-youtube"></i>
          </div>
        </div>
      </div>

      <div className="p-5 flex justify-end bg-footerBlue">
        <p>
          Built by
          <a
            href="https://github.com/erinsophie"
            aria-label="Link to GitHub page"
            target="_blank"
            rel="noopener noreferrer"
            className="hover:underline"
          >
            {' '}
            erinsophie
          </a>{' '}
          |{' '}
          <a
            href="https://github.com/erinsophie/shopping-cart"
            aria-label="Link to GitHub repository"
            target="_blank"
            rel="noopener noreferrer"
            className="hover:underline"
          >
            {' '}
            Source{' '}
          </a>
        </p>
      </div>
    </footer>
  );
}

export default Footer;
