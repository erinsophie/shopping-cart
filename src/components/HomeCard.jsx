import { Link } from 'react-router-dom';
import mug from '../assets/products/mug.jpg';

function HomeCard() {
  return (
    <div className="flex p-8 gap-6 items-center justify-between border border-gray-400 rounded-xl md:gap-12 md:flex-row md:p-10 lg:gap-12 lg:flex-row lg:p-14">
      <div className="flex flex-col gap-2 lg:gap-10">
        <div className="flex flex-col gap-2.5">
          <p className="font-light text-sm text-gray-500">
            Editor&apos;s Picks
          </p>
          <p className="text-base md:text-xl lg:text-2xl">Home Decor</p>
          <Link to="/2">
            <p className="cursor-pointer text-sm md:text-base lg:text-base">
              Shop these unqiue finds{' '}
              <i className="fa-solid fa-arrow-right"></i>
            </p>
          </Link>
        </div>

        <div className="flex flex-col hidden gap-4 text-xs lg:text-base md:block lg:block">
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat.
          </p>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim
            ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut
            aliquip ex ea commodo consequat. Duis aute irure dolor in
            reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
            pariatur.
          </p>
        </div>
      </div>
      <img
        src={mug}
        alt="ceramic mugs"
        className="w-44 rounded-xl cursor-pointer hover:shadow-xl md:w-44 lg:w-64"
      ></img>
    </div>
  );
}

export default HomeCard;
