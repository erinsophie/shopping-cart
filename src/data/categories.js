import backToSchool from '../assets/categories/back-to-school.jpg';
import personalisedGifts from '../assets/categories/personalised-gifts.jpg';
import jewellery from '../assets/categories/jewellery.jpg';
import homeDecor from '../assets/categories/home-decor.jpg';

const categories = [
  {
    id: 1,
    name: 'Back to school',
    image: backToSchool,
  },
  {
    id: 2,
    name: 'Home decor',
    image: homeDecor,
  },
  {
    id: 3,
    name: 'Jewellery',
    image: jewellery,
  },
  {
    id: 4,
    name: 'Personalised gifts',
    image: personalisedGifts,
  },
];

export default categories;
