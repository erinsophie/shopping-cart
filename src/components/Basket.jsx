import { useBasket } from './BasketContext';
import { Link } from 'react-router-dom';

function Basket() {
  const {
    basket,
    showBasket,
    setShowBasket,
    getSubtotal,
    getBasketAmount,
    deleteProduct,
    clearBasket,
    updateQuantity,
  } = useBasket();

  return showBasket ? (
    <div className="fixed top-0 right-0 z-10 w-full h-full bg-white p-8 overflow-y-auto border border-black flex flex-col gap-10 md:w-1/2 lg:w-1/3">
      <div className="flex justify-between">
        <div className="flex flex-col gap-2">
          <p>Your basket</p>
          <p data-testid="items-amount">{getBasketAmount()} items</p>
        </div>
        <button
          onClick={() => setShowBasket((prev) => !prev)}
          className="fa-solid fa-xmark text-2xl transform transition-transform ease-in-out duration-200 hover:scale-110"
          data-testid="close-basket-btn"
        ></button>
      </div>

      {getBasketAmount() === 0 ? (
        <div className="flex justify-center">
          <p className="font-cormorant text-2xl">Basket is empty</p>
        </div>
      ) : (
        <div className="flex flex-col gap-4">
          {basket.map((basketItem) => (
            // each basket item
            <div
              key={basketItem.product.id}
              className="flex border border-black"
              data-testid="basket-item"
            >
              <div className="w-32 h-36">
                <img
                  src={basketItem.product.image}
                  alt={basketItem.product.title}
                  className="w-full h-full object-cover"
                ></img>
              </div>

              <div className="flex flex-col justify-between w-full p-4">
                <div className="flex justify-between">
                  <p>{basketItem.product.title}</p>
                  <button
                    onClick={() => deleteProduct(basketItem.product.id)}
                    className="fa-regular fa-trash-can"
                    data-testid={`delete-button-${basketItem.product.id}`}
                  ></button>
                </div>
                <div className="flex justify-between">
                  <p>£{basketItem.product.price}</p>
                  <div className="flex gap-3">
                    <button
                      onClick={(e) => updateQuantity(basketItem, e)}
                      className="bg-gray-300 w-7"
                    >
                      -
                    </button>
                    <p>{basketItem.quantity}</p>
                    <button
                      onClick={(e) => updateQuantity(basketItem, e)}
                      className="bg-gray-300 w-7"
                    >
                      +
                    </button>
                  </div>
                </div>
              </div>
            </div>
          ))}
        </div>
      )}

      <div className="flex justify-between items-center">
        <p data-testid="subtotal">Subtotal: £{getSubtotal(basket)}</p>
        {getBasketAmount() > 0 ? (
          <Link to="/checkout">
            <button
              onClick={clearBasket}
              className="border border-black p-2.5 transform transition-transform ease-in-out duration-200 hover:scale-110"
            >
              Checkout
            </button>
          </Link>
        ) : null}
      </div>
    </div>
  ) : null;
}

export default Basket;
