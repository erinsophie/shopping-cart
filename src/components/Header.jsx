import { useBasket } from '../components/BasketContext';
import { Link } from 'react-router-dom';

function Header() {
  const { showBasket, setShowBasket, getBasketAmount } = useBasket();

  return (
    <header className="bg-purple text-white flex items-center gap-3 p-3 border-b border-gray-300 md:p-6 md:gap-6 md:pl-16 md:pr-16 lg:pl-28 lg:pr-28 lg:p-5 lg:gap-10">
      <div className="flex items-center gap-3 flex-1">
        <Link to="/">
          <h1 className="text-xl md:text-2xl lg:text-3xl">FakeStore</h1>
        </Link>

        <form className="w-full flex gap-3">
          <label htmlFor="searchInput" className="sr-only">
            Search Query
          </label>
          <input
            id="searchInput"
            type="search"
            placeholder="Search for anything"
            className="text-black rounded-full bg-gray-100 w-full p-1 font-light text-xs md:text-sm lg:p-2.5 lg:text-base"
          />
          <button type="submit" aria-label="Search">
            <i className="fa-solid fa-magnifying-glass text-sm lg:text-xl"></i>
          </button>
        </form>
      </div>

      <div className="flex gap-2 items-center transform transition-transform ease-in-out duration-200 hover:scale-110">
        <button
          aria-label="Shopping basket"
          aria-expanded={showBasket}
          data-testid="basket-icon"
          onClick={() => setShowBasket((prev) => !prev)}
        >
          <i className="fa-solid fa-basket-shopping text-sm lg:text-xl"></i>
        </button>

        <p data-testid="basket-quantity">{getBasketAmount()}</p>
      </div>
    </header>
  );
}

export default Header;
