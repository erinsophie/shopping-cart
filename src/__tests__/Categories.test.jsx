import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { BasketProvider } from '../components/BasketContext';
import Categories from '../components/Categories';
import categories from '../data/categories';

describe('Render tests', () => {
  beforeEach(() => {
    render(
      <MemoryRouter>
        <BasketProvider>
          <Categories />
        </BasketProvider>
      </MemoryRouter>,
    );
  });

  it('Renders the main tagline correctly', () => {
    const mainText = screen.getByText(
      'An online marketplace for creative crafts!',
    );
    expect(mainText).toBeInTheDocument();
  });

  it('Renders correct amount of links', () => {
    const links = screen.getAllByRole('link');
    expect(links.length).toBe(categories.length);
  });

  it('Each Link has correct href attribute', () => {
    categories.forEach((category) => {
      const link = screen.getByRole('link', { name: category.name });
      expect(link).toHaveAttribute('href', `/${category.id}`);
    });
  });

  it('Passes the correct props to CategoryPreview', () => {
    // mock category preview page
    vi.mock('../components/CategoryPreview', () => {
      return {
        default: function MockCategoryPreview({ src, name }) {
          return (
            <div
              data-testid="mock-category-preview"
              data-testsrc={src}
              data-testname={name}
            >
              <img src={src}></img>
              <p>{name}</p>
            </div>
          );
        },
      };
    });

    render(
      <MemoryRouter>
        <Categories />
      </MemoryRouter>,
    );
    const previews = screen.getAllByTestId('mock-category-preview');

    categories.forEach((category, index) => {
      const mockPreview = previews[index];
      expect(mockPreview.getAttribute('data-testsrc')).toBe(category.image);
      expect(mockPreview.getAttribute('data-testname')).toBe(category.name);
    });
  });
});
