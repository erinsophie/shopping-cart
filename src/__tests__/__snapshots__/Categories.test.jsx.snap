// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render tests > Matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center gap-7 p-5 bg-pinkBeige md:p-8 lg:p-8"
  >
    <h2
      class="font-cormorant text-xl md:text-2xl lg:text-4xl"
    >
      An online marketplace for creative crafts!
    </h2>
    <div
      class="flex gap-5 md:gap-16 lg:gap-16"
    >
      <a
        aria-label="Back to school"
        href="/1"
      >
        <div
          data-testid="mock-category-preview"
        >
          <img
            src="/src/assets/categories/back-to-school.jpg"
          />
          <p>
            Back to school
          </p>
        </div>
      </a>
      <a
        aria-label="Home decor"
        href="/2"
      >
        <div
          data-testid="mock-category-preview"
        >
          <img
            src="/src/assets/categories/home-decor.jpg"
          />
          <p>
            Home decor
          </p>
        </div>
      </a>
      <a
        aria-label="Jewellery"
        href="/3"
      >
        <div
          data-testid="mock-category-preview"
        >
          <img
            src="/src/assets/categories/jewellery.jpg"
          />
          <p>
            Jewellery
          </p>
        </div>
      </a>
      <a
        aria-label="Clothing"
        href="/4"
      >
        <div
          data-testid="mock-category-preview"
        >
          <img
            src="/src/assets/categories/clothing.jpg"
          />
          <p>
            Clothing
          </p>
        </div>
      </a>
    </div>
  </div>
</div>
`;
